README
This APS system is copyright John Boyland 1996-2019.
You may freely use this software provided you don't claim authorship.

```sh
docker run -it amir734jj/aps bash
> cd /usr/lib/aps/examples
> make first.scc.sched
```

history.txt
    1  ls
    2  cd ~/cs655
    3  l
    4  sls
    5  klog
    6  exit
    7  cd ~/
    8  ls
    9  exit
   10  pico ~/.zshrc
   11  exit
   12  pico ~/.zshrc
   13  exit
   14  ls
   15  klog
   16  pico ~/.zshrc
   17  pico ~/.bashrc
   18  ls
   19  pico ~/.zshrc
   20  pico ~/.bashrc
   21  gedit ~/.bashrc
   22  open .
   23  gedit ~/.zshrc
   24  gedit ~/.zshrc &
   25  more ~.zshenv
   26  more ~/.zshenv
   27  gedit ~/.zshenv
   28  klog
   29  pico ~/.zshenv
   30  andrew
   31  ls
   32  pico ~/.bashrc
   33  gedit ~/.bashrc
   34  clear
   35  ls
   36  zshell
   37  zsh
   38  ls
   39  klog
   40  ls
   41  cd ~/cs655
   42  ls
   43  cd private
   44  ls
   45  cd cmd
   46  ls
   47  ./on-deadline
   48  ./on-deadline PA3
   49  clear
   50  ls
   51  ./enter-short-grades
   52  python status.py
   53  python status.py 801 lab1
   54  python status.py 801 lab2
   55  python status.py 801 lab3
   56  python status.py 801 lab4
   57  python status.py 801 lab5
   58  python status.py 801 lab6
   59  python status.py 801 lab7
   60  python status.py 801 lab8
   61  python status.py 801 lab7
   62  ./enter-short-grades 
   63  ./enter-short-grades lab7
   64  ./enter-short-grades lab8
   65  python status.py
   66  python status.py lab5
   67  ls
   68  ./make-grade-summary
   69  ./make-grade-summary 801
   70  ls
   71  cd ../
   72  ls
   73  cd private/assignment/5
   74  ls
   75  more README
   76  clear
   77  ls
   78  make
   79  ls
   80  more M
   81  more Makefile
   82  make PA5.jar
   83  clear
   84  ls
   85  make
   86  make clean
   87  ls
   88  make
   89  cd /afs/cs.uwm.edu/project/cool/scala/PA5.jar
   90  cd /afs/cs.uwm.edu/project/cool/scala/
   91  ls
   92  ls *.jar
   93  cd ../
   94  ls *.jar
   95  ls
   96  ack PA5.jar
   97  awk PA5.jar
   98  open .
   99  ls
  100  lsfind . -name "PA4.jar" -print
  101  find . -name "PA4.jar" -print
  102  open .
  103  ls
  104  cd scala
  105  make
  106  ls
  107  more Ma
  108  more Makefile
  109  make PA5.jar
  110  ls
  111  ls *.jar
  112  pwd
  113  make PA4.jar
  114  gedit Makefile
  115  make PA4.jar
  116  andrew
  117  klog
  118  ls
  119  pwd
  120  ls *.jar
  121  cp PA5.jar ~/cs655/private/assignment/5/
  122  cd ~/cs655/private/assignment/5
  123  ls
  124  make
  125  ls
  126  more Semant.scala
  127  gedit Semant.scala
  128  make src.install
  129  ls
  130  cd ../
  131  open .
  132  cd /afs/cs.uwm.edu/project/cool/scala/
  133  ls
  134  cd ../
  135  open .
  136  cd ~/cs655/
  137  find . -name good.cool
  138  cd src
  139  ls
  140  cd PA5
  141  ls
  142  cd /801/cjradtke/OldFiles/PA4
  143  cd ~/cs655/801/cjradtke
  144  cd PA4
  145  ls
  146  more good.cool
  147  more bad.cool
  148  ls *.cool
  149  cp good.cool ~/cs655/private/assignment/5/
  150  cp bad.cool ~/cs655/private/assignment/5/
  151  ls
  152  cd ~/cs655/private/assignment
  153  ls
  154  cd 5
  155  make src.install
  156  make 
  157  cd ../
  158  cd src
  159  ls
  160  cd PA5
  161  ls
  162  more good.cool.SKEL
  163  more bad.cool.SKEL
  164  cd ../
  165  cd private/assignment/5
  166  ls
  167  make asn5.install
  168  andrew
  169  cd ../
  170  cd www
  171  pico index.html
  172  vim index.html
  173  cd ../
  174  cd src
  175  cd PA5
  176  ls
  177  gedit Semant.scala.SKEL
  178  cd ../
  179  cd cmd
  180  python status.py
  181  python status.py 801 
  182  python status.py 801 PA3
  183  python status.py 802 PA3
  184  python status.py 802 PA2
  185  python status.py 802 PA1
  186  cd ../
  187  ls
  188  cd 8001
  189  ls
  190  cd 800
  191  ls
  192  cd hesamian
  193  ls
  194  cd labs
  195  ls
  196  more 3
  197  more 4
  198  ls
  199  clear
  200  ls
  201  pwd
  202  whoami
  203  klog
  204  ls
  205  pwd
  206  cd ../
  207  klog
  208  cd ../
  209  ls
  210  cd ../
  211  cd 801
  212  ls
  213  cd cjradtke
  214  ls
  215  cd PA4
  216  ls
  217  make test
  218  more Makefile
  219  clear
  220  ls
  221  make diff
  222  make
  223  make test
  224  more Semant.scala
  225  clear
  226  ls
  227  more good.cool
  228  more README
  229  cd ../
  230  ls
  231  cd oconnori
  232  ls
  233  cd PA4
  234  ls
  235  make 
  236  make test
  237  ls
  238  more Makefile
  239  ls
  240  more Semant.scala
  241  ls
  242  cd ../
  243  ls
  244  cd private/assignment/4
  245  ls
  246  more grade-script
  247  ls
  248  cd ../3
  249  ls
  250  more old-grading-sheet
  251  open .
  252  cd ~/cool
  253  ls
  254  find . -name "grading-sheet"
  255  ls
  256  clear
  257  ls
  258  cd OldFiles
  259  ls
  260  cd ../
  261  cd solutions/
  262  ls
  263  cd PA4
  264  ls
  265  more grading-sheet
  266  cp grading-sheet ~/cs655/private/assignment/5
  267  cd ~/cs655/private/assignment/4
  268  ls
  269  mv  ../5/grading-sheet .
  270  ls
  271  gedit grading-sheet
  272  ls
  273  cd ../
  274  cd 800
  275  ls
  276  rm -rf temp temp2 temp3
  277  ls
  278  cd fixed
  279  ls
  280  cd ../
  281  rm -rf fixed
  282  ls
  283  clear
  284  ls
  285  cd hesamian
  286  ls
  287  rm Cool.y
  288  ls
  289  mkdir grading
  290  cd grading
  291  make -f /afs/cs.uwm.edu/users/classes/cs655/src/PA4/Makefile
  292  ls
  293  make test
  294  make 
  295  more Makefile
  296  make test
  297  cp good.cool ../../../801/weinertw/PA4/good.cool
  298  cp ../../../801/weinertw/PA4/good.cool .
  299  cp ../../../801/weinertw/PA4/bad.cool .
  300  make test
  301  ls
  302  pico .gitignore
  303  ls
  304  git status
  305  git add .
  306  git status
  307  git commit -m "started work"
  308  git push
  309  git remote add origin https://github.com/amir734jj/hesamian-chat-room.git
  310  git push
  311  git push --set-upstream origin master
  312  git p
  313  klog
  314  pwd
  315  ls
  316  cd ~/cs655
  317  ls
  318  cd private/cmd
  319  ./on-deadline
  320  cd ~/my-afs
  321  ls
  322  cd aps-clone
  323  pwd
  324  code .
  325  ./on-deadline PA4
  326  pico ~/.zshrc
  327  man zshmisc

analyze/aps-info.h
/* APS-INFO
 * Access slots inside APS nodes.
 * This file must be modified and aps-info.c recompiled each time
 * we need a new slot.
 */

struct Program_info {
  unsigned program_flags;
#define PROGRAM_BOUND_FLAG 1
#define PROGRAM_TYPED_FLAG 2
};
extern struct Program_info *Program_info(Program);
#define PROGRAM_IS_BOUND(p) (Program_info(p)->program_flags&PROGRAM_BOUND_FLAG)
#define PROGRAM_IS_TYPED(p) (Program_info(p)->program_flags&PROGRAM_TYPED_FLAG)

struct Unit_info {
  Unit next_unit;
#define UNIT_NEXT(u) Unit_info(u)->next_unit
};
extern struct Unit_info *Unit_info(Unit);

struct Use_info {
  Declaration use_decl;
  TypeEnvironment use_type_env;
};
extern struct Use_info *Use_info(Use);
#define USE_DECL(u) (Use_info(u)->use_decl)
#define USE_TYPE_ENV(u) (Use_info(u)->use_type_env)

struct Declaration_info {
  Declaration next_decl; /* in a sequence of Declarations */
  Declaration next_field_decl; /* next field (init'ed in aps-fiber) */
  Declaration dual_decl; /* a declaration created out of nowhere */
  Declaration copied_decl; /* pointer to copied version (scratch) */
  FIBERSETS decl_fibersets;
  int if_index; /* count of if_stmt within a top-level match */
#define instance_index if_index
  CONDITION decl_cond; /* condition that must be satisfied to take effect */
  struct summary_dependency_graph *node_phy_graph;
  unsigned decl_flags;
  void * call_sites;
  void *analysis_state;
#define DECL_LHS_FLAG 1
#define DECL_RHS_FLAG 2
#define DECL_OBJECT_FLAG 4
#define TYPE_FORMAL_EXTENSION_FLAG 8
#define ATTR_DECL_SYN_FLAG 16
#define ATTR_DECL_INH_FLAG 32
#define FIELD_DECL_STRICT_FLAG 64
#define FIELD_DECL_CYCLIC_FLAG 128
#define FIELD_DECL_FLAG 256
#define DECL_LOCAL_FLAG 512 /* declaration declared in module being analyzed */
#define FIELD_DECL_REVERSE_FLAG 1024
#define SHARED_INFO_FLAG (1<<11)
#define SHARED_DECL_FLAG (1<<12) /* decl'n declared at top-level of module */
#define START_PHYLUM_FLAG (1<<13)
#define FIELD_DECL_UNTRACKED_FLAG (1<<14)
#define SELF_MANAGED_FLAG (1<<15)
#define UP_DOWN_FLAG (1<<16)
#define MODULE_DECL_GENERATING (1<<17)
#define MODULE_DECL_GENERATING_VALID (1<<18)
	OSET oset;				/* oset of the declaration */
	USET uset;				/* uset of the declaration */
	int	 index;			  /* represent the nodeis Qd and Qd(-) */
										/* odd: Qd, even: Qd(-). such as (1,2) */
};
extern struct Declaration_info *Declaration_info(Declaration);

#define DECL_NEXT(decl) (Declaration_info(decl)->next_decl)
#define NEXT_FIELD(decl) (Declaration_info(decl)->next_field_decl)
#define DUAL_DECL(decl) (Declaration_info(decl)->dual_decl)
#define MODULE_SHARED_INFO_PHYLUM(m) DUAL_DECL(m)
#define MODULE_SHARED_INFO_POLYMORPHIC(m) NEXT_FIELD(m)
#define POLYMORPHIC_INSTANCES(a) DUAL_DECL(a) /* not used generally yet */
#define COPIED_DECL(decl) (Declaration_info(decl)->copied_decl)
#define DECL_IS_LHS(decl) (Declaration_info(decl)->decl_flags&DECL_LHS_FLAG)
#define DECL_IS_RHS(decl) (Declaration_info(decl)->decl_flags&DECL_RHS_FLAG)
#define DECL_IS_OBJECT(decl) \
  (Declaration_info(decl)->decl_flags&DECL_OBJECT_FLAG)
#define DECL_IS_SYNTAX(decl) \
  (Declaration_info(decl)->decl_flags&(DECL_LHS_FLAG|DECL_RHS_FLAG))
#define TYPE_FORMAL_IS_EXTENSION(decl) \
  (Declaration_info(decl)->decl_flags&TYPE_FORMAL_EXTENSION_FLAG)
#define ATTR_DECL_IS_SYN(decl) \
  (Declaration_info(decl)->decl_flags&ATTR_DECL_SYN_FLAG)
#define ATTR_DECL_IS_INH(decl) \
  (Declaration_info(decl)->decl_flags&ATTR_DECL_INH_FLAG)
#define FIELD_DECL_IS_STRICT(decl) \
  (Declaration_info(decl)->decl_flags&FIELD_DECL_STRICT_FLAG)
#define FIELD_DECL_IS_UNTRACKED(decl) \
  (Declaration_info(decl)->decl_flags&FIELD_DECL_UNTRACKED_FLAG)
#define FIELD_DECL_IS_CYCLIC(decl) \
  (Declaration_info(decl)->decl_flags&FIELD_DECL_CYCLIC_FLAG)
#define FIELD_DECL_P(decl) \
  (Declaration_info(decl)->decl_flags&FIELD_DECL_FLAG)
#define DECL_IS_LOCAL(decl) \
  (Declaration_info(decl)->decl_flags&DECL_LOCAL_FLAG)
#define FIELD_DECL_IS_REVERSE(decl) \
  (Declaration_info(decl)->decl_flags&FIELD_DECL_REVERSE_FLAG)
#define ATTR_DECL_IS_SHARED_INFO(decl) \
  (Declaration_info(decl)->decl_flags&SHARED_INFO_FLAG)
#define ATTR_DECL_IS_UP_DOWN(decl) \
  (Declaration_info(decl)->decl_flags&UP_DOWN_FLAG)
#define DECL_IS_SHARED(decl) \
  (Declaration_info(decl)->decl_flags&SHARED_DECL_FLAG)
#define DECL_IS_START_PHYLUM(decl) \
  (Declaration_info(decl)->decl_flags&START_PHYLUM_FLAG)
#define DECL_IS_SELF_MANAGED(decl) \
  (Declaration_info(decl)->decl_flags&SELF_MANAGED_FLAG)

#define fibersets_for(decl) (Declaration_info(decl)->decl_fibersets)
#define fiberset_for(decl,fstype) (fibersets_index(fibersets_for(decl),fstype))

struct Match_info {
  Declaration header; /* for, case or top-level-match */
  Match next_match;
  int if_index; /* count of if_stmt within a top-level match */
  CONDITION match_cond; /* condition that must be satisfied to take effect */
  Expression match_test; /* last expression that must be evaluated,
			   expr_next points to previous, back to first */
};
extern struct Match_info *Match_info(Match);
#define MATCH_NEXT(m) Match_info(m)->next_match
  
struct Expression_info {
  Expression next_actual; /* in a sequence of args to a function call */
#define next_expr next_actual
  Type expr_type; /* current not computed or used */
  Declaration call_decl; /* attribute or function decl called here */
  FIBERSETS expr_fibersets;
  int expr_flags;
  int expr_helper_num; /* used in incremental impl. */
  struct attribute_instance* value_for; /* This is the RHS that gives value to this instance */
  Declaration funcall_proxy; /* This is a Decl that permits a funcall node to be an INSTANCE node */
#define EXPR_LHS_FLAG 1
#define EXPR_RHS_FLAG 2
	int	 index;			// represent the nodes of Qe and Qe(-).
};
extern struct Expression_info *Expression_info(Expression);
#define EXPR_NEXT(expr) (Expression_info(expr)->next_expr)
#define EXPR_IS_LHS(expr) (Expression_info(expr)->expr_flags&EXPR_LHS_FLAG)
#define EXPR_IS_RHS(expr) (Expression_info(expr)->expr_flags&EXPR_RHS_FLAG)

#define expr_fibersets_for(expr) (Expression_info(expr)->expr_fibersets)
#define expr_fiberset_for(expr,fstype) \
  (fibersets_index(expr_fibersets_for(expr),fstype))
#define expr_helper_for(expr) (Expression_info(expr)->expr_helper_num)

struct Pattern_info {
  Pattern next_pattern_actual; /* in a sequence of args to a pattern call */
#define next_pattern next_pattern_actual
  Type pat_type;
  void* pat_impl; /* how it is implemented */
};
extern struct Pattern_info *Pattern_info(Pattern);
#define PAT_NEXT(pat) (Pattern_info(pat)->next_pattern_actual)

struct Type_info {
  Type next_type_actual; /* in a series of type actuals */
  void *binding_temporary;
  char *impl_type;
  Signature  *type_sig;
};
extern struct Type_info *Type_info(Type);
#define TYPE_NEXT(type) (Type_info(type)->next_type_actual)

struct Signature_info {
  void *binding_temporary;
};
extern struct Signature_info *Signature_info(Signature);

struct Def_info {
  int unique_prefix; /* to distinguish local attributes */
};
extern struct Def_info *Def_info(Def);

extern void set_tnode_parent(Program p);

extern void *tnode_parent(void *);

#define decl_name(decl) (char*)symbol_name(def_name(declaration_def(decl)))

/* TypeActuals utility */
int compute_type_contour_size(TypeActuals type_actuals, Declarations type_formals);